substrate/runtime/democracy/src/vote_threshold.rs

```diff
@@ -58,18 +58,64 @@ pub trait Approved<Balance> {
 	fn approved(&self, approve: Balance, against: Balance, electorate: Balance) -> bool;
 }
 
-impl<Balance: IntegerSquareRoot + Ord + Add<Balance, Output = Balance> + Mul<Balance, Output = Balance> + Div<Balance, Output = Balance> + Copy> Approved<Balance> for VoteThreshold {
+/// Compare two fractions.
+pub trait CompareRationals {
+	/// Return `true` iff `n1 / d1 < n2 / d2`. `d1` and `d2` may not be zero.
+	fn compare_rationals(n1: Self, d1: Self, n2: Self, d2: Self) -> bool;
+}
+
+impl<T: Zero + Mul<T, Output = T> + Div<T, Output = T> + Rem<T, Output = T> + Ord + Copy> CompareRationals for T {
+	fn compare_rationals(n1: Self, d1: Self, n2: Self, d2: Self) -> bool {
+		// Uses a continued fractional representation for a non-overflowing compare.
+		// Detailed at https://janmr.com/blog/2014/05/comparing-rational-numbers-without-overflow/.
+		let q1 = n1 / d1;
+		let q2 = n2 / d2;
+		if q1 < q2 {
+			return true;
+		}
+		if q2 < q1 {
+			return false;
+		}
+		let r1 = n1 % d1;
+		let r2 = n2 % d2;
+		if r2.is_zero() {
+			return false;
+		}
+		if r1.is_zero() {
+			return true;
+		}
+		// AUDIT: could be dangerous if no tail recursion optimisation as votes could be crafted
+		// to overflow the stack.
+		Self::compare_rationals(d2, r2, d1, r1)
```

Comment:
maybe use a loop then? AFAIK Rust doesn't support TCO in many cases or perhaps at all and there have been no successful RFCs for guaranteed TCO, so it's better not to be at the mercy of the compiler.